src/media_devices/mod.rs

```diff
@@ -0,0 +1,124 @@
+pub enum MediaDeviceInfoKind {
+    VideoInput,
+    AudioInput,
+    AudioOutput,
+}
+
+/// Describes a single media input or output device
+///
+/// Call [`enumerate_devices`] to obtain a list of devices for your hardware.
+#[derive(Debug)]
+pub struct MediaDeviceInfo {
+    device_id: String,
+    group_id: Option<String>,
+    kind: MediaDeviceInfoKind,
+    label: String,
+    device: Box<dyn std::any::Any>,
+}
+
+impl MediaDeviceInfo {
+    pub(crate) fn new(
+        device_id: String,
+        group_id: Option<String>,
+        kind: MediaDeviceInfoKind,
+        label: String,
+        device: Box<dyn std::any::Any>,
+    ) -> Self {
+        Self {
+            device_id,
+            group_id,
+            kind,
+            label,
+            device,
+        }
+    }
+
+    /// Identifier for the represented device
+    ///
+    /// The current implementation is not stable across sessions so you should not persist this
+    /// value
+    pub fn device_id(&self) -> &str {
+        &self.device_id
+    }
+
+    /// Two devices have the same group identifier if they belong to the same physical device
+    pub fn group_id(&self) -> Option<&str> {
+        self.group_id.as_deref()
+    }
+
+    /// Enumerated value that is either "videoinput", "audioinput" or "audiooutput".
+    pub fn kind(&self) -> MediaDeviceInfoKind {
+        self.kind
+    }
+
+    /// Friendly label describing this device
+    pub fn label(&self) -> &str {
+        &self.label
+    }
+
+    pub(crate) fn device(self) -> Box<dyn std::any::Any> {
+        self.device
+    }
+}
+
+/// Prompt for permission to use a media input (audio only)
+///
+/// This produces a [`MediaStream`] with tracks containing the requested types of media, which can
+/// be used inside a [`MediaStreamAudioSourceNode`](crate::node::MediaStreamAudioSourceNode).
+///
+/// It is okay for the `MediaStream` struct to go out of scope, any corresponding stream will still be
+/// kept alive and emit audio buffers. Call the `close()` method if you want to stop the media
+/// input and release all system resources.
+///
+/// # Example
+///
+/// ```no_run
+/// use web_audio_api::context::{BaseAudioContext, AudioContext};
+/// use web_audio_api::context::{AudioContextLatencyCategory, AudioContextOptions};
+/// use web_audio_api::media_devices;
+/// use web_audio_api::node::AudioNode;
+///
+/// let context = AudioContext::default();
+/// let mic = media_devices::get_user_media();
+///
+/// // register as media element in the audio context
+/// let background = context.create_media_stream_source(&mic);
+///
+/// // connect the node directly to the destination node (speakers)
+/// background.connect(&context.destination());
+///
+/// // enjoy listening
+/// std::thread::sleep(std::time::Duration::from_secs(4));
+/// ```
+// TODO, return Promise? How to provide constraints?
+pub fn get_user_media() -> MediaStream {
+    crate::io::build_input(AudioContextOptions::default())
```

Comment:
Don't really know what would be the proper API, but being able to pass the `deviceId` somehow would be very important (e.g. to work to with Jack or any external sound cards).

I could just re-wrap the whole thing on the Node.js side, but as everything is now `pub(crate)` I will be stuck

Reply:
maybe something like this (a bit verbose...):

```rs
#[non_exhaustive]
struct MediaStreamConstraints {
    audio: MediaTrackConstraintsOption
}

#derive(Default)
enum MediaTrackConstraintsOption {
   False: boolean,
   #[default] // not spec compliant, but makes sens in our case I guess
   True: boolean,
   MediaTrackConstraints: MediaTrackConstraints
}

#[non_exhaustive]
struct MediaTrackConstraints {
      deviceId: string
}
```

Then in the general case, we could just:
```rs
let stream = media_devices.get_user_media(Default::default())
```